File: lc/coreprep/implementation/practice_iterations/merge_intervals_5/ConflictingAppointments.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(nlogn) / O(1)
public class ConflictingAppointments {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  private static boolean hasConflict(Interval a, Interval b) {
    return (a.start >= b.start && a.start < b.end) ||
        (b.start >= a.start && b.start < a.end);
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>();
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        a.sort((x, y) -> Integer.compare(x.start, y.start));

        boolean canAttendAll = true;
        for (int i = 1; i < n; i++) {
          if (hasConflict(a.get(i), a.get(i - 1))) {
            canAttendAll = false;
            break;
          }
        }

        System.out.println(canAttendAll);
      }
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/conflicting_appointments.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval() : start(-1), end(-1) {}

  Interval(int start, int end) : start(start), end(end) {}
};

// O(nlogn) / O(1)
bool can_attend_all(vector<Interval> &a) {
  sort(a.begin(), a.end(),
       [](const auto &e1, const auto &e2) { return e1.start <= e2.start; });

  auto has_overlap = [](const auto &e1, const auto &e2) {
    return (e1.start >= e2.start && e1.start < e2.end) ||
           (e2.start >= e1.start && e2.start < e1.end);
  };

  for (int i = 0; i < a.size() - 1; i++) {
    if (has_overlap(a[i], a[i + 1])) {
      return false;
    }
  }

  return true;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, start, end;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a(n);
    for (int i = 0; i < n; i++) {
      cin >> start >> end;
      a[i] = Interval(start, end);
    }

    cout << (can_attend_all(a) ? "true" : "false") << "\n";
  }

  return 0;
}